pkg/tracker/reward/reward.go

```diff
@@ -0,0 +1,412 @@
+	ctx              context.Context
+	stop             context.CancelFunc
+	addr             common.Address
+
+	tsDB   *tsdb.DB
+	aggr   aggregator.IAggregator
+	engine *promql.Engine
+}
+
+func NewRewardTracker(
+	logger log.Logger,
+	ctx context.Context,
+	cfg Config,
+	tsDB *tsdb.DB,
+	client *ethclient.Client,
+	contractInstance *contracts.ITellor,
+	addr common.Address,
+	aggr aggregator.IAggregator,
+) (*RewardTracker, error) {
+	logger, err := logging.ApplyFilter(cfg.LogLevel, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "apply filter logger")
+	}
+	logger = log.With(logger, "component", ComponentName)
+
+	abi, err := abi.JSON(strings.NewReader(tellor.TellorABI))
+	if err != nil {
+		return nil, errors.Wrap(err, "abi read")
+	}
+
+	opts := promql.EngineOpts{
+		Logger:               logger,
+		Reg:                  nil,
+		MaxSamples:           30000,
+		Timeout:              10 * time.Second,
+		LookbackDelta:        5 * time.Minute,
+		EnableAtModifier:     true,
+		EnableNegativeOffset: true,
+	}
+	engine := promql.NewEngine(opts)
+
+	ctx, cncl := context.WithCancel(ctx)
+	return &RewardTracker{
+		client:           client,
+		logger:           logger,
+		contractInstance: contractInstance,
+		abi:              abi,
+		addr:             addr,
+		ctx:              ctx,
+		stop:             cncl,
+		tsDB:             tsDB,
+		engine:           engine,
+		aggr:             aggr,
+	}, nil
+}
+
+func (self *RewardTracker) Start() error {
+	level.Info(self.logger).Log("msg", "starting")
+
+	go self.monitorGas()
+
+	<-self.ctx.Done()
+	return nil
+}
+
+func (self *RewardTracker) Stop() {
+	self.stop()
+}
+
+func (self *RewardTracker) monitorGas() {
+	var err error
+	ticker := time.NewTicker(DefaultRetry)
+	defer ticker.Stop()
+
+	logger := log.With(self.logger, "event", "NonceSubmitted")
+
+	var sub event.Subscription
+	events := make(chan *tellor.TellorNonceSubmitted)
+	for {
+		select {
+		case <-self.ctx.Done():
+			return
+		default:
+		}
+		sub, err = self.nonceSubmittedSub(events)
+		if err != nil {
+			level.Error(logger).Log("msg", "initial subscribing to events failed")
+			<-ticker.C
+			continue
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-self.ctx.Done():
+			return
+		case err := <-sub.Err():
+			if err != nil {
+				level.Error(logger).Log(
+					"msg",
+					"subscription error",
+					"err", err)
+			}
+
+			// Trying to resubscribe until it succeeds.
+			for {
+				select {
+				case <-self.ctx.Done():
+					return
+				default:
+				}
+				sub, err = self.nonceSubmittedSub(events)
+				if err != nil {
+					level.Error(logger).Log("msg", "re-subscribing to events failed", "err", err)
+					<-ticker.C
+					continue
+				}
+				break
+			}
+			level.Info(logger).Log("msg", "re-subscribed to events")
+		case event := <-events:
+			err := self.recordGasUsage(event)
+			if err != nil {
+				level.Error(self.logger).Log("msg", "record gas usage", "err", err)
+			}
+			err = self.recordGasUsageEstimated(self.ctx)
```

Comment:
don't need to pass context here as can get it directly in `recordGasUsageEstimated`
```suggestion
			err = self.recordGasUsageEstimated()
```